javaslang/src/main/java/javaslang/concurrent/Future.java

```diff
@@ -309,10 +309,45 @@
      * @throws NullPointerException if one of executorService of computation is null.
      */
     static <T> Future<T> of(ExecutorService executorService, CheckedSupplier<? extends T> computation) {
+        return Future.of(executorService, computation, 0);
+    }
+
+    /**
+     * Starts an asynchronous computation after delaying, backed by the {@link #DEFAULT_EXECUTOR_SERVICE}.
+     * The thread supplied by the {@code DEFAULT_EXECUTOR_SERVICE} will be blocked at least {@code delay} milliseconds.
+     * If the {@code delay} &lt;= 0, the {@code computation} will be executed immediately.
+     *
+     * @param computation A computation.
+     * @param delay    The time in milliseconds to wait before executing the given {@code computation}.
+     * @param <T>         Type of the computation result.
+     * @return A new Future instance.
+     * @throws NullPointerException if computation is null.
+     */
+    static <T> Future<T> of(CheckedSupplier<? extends T> computation, long delay) {
+        return Future.of(DEFAULT_EXECUTOR_SERVICE, computation, delay);
+    }
+
+    /**
+     * Starts an asynchronous computation after delaying, backed by the given {@link ExecutorService}.
+     * The thread supplied by the {@code ExecutorService} will be blocked at least {@code delay} milliseconds.
+     * If the {@code delay} &lt;= 0, the {@code computation} will be executed immediately.
+     *
+     * @param executorService An executor service.
+     * @param computation     A computation.
+     * @param delay        The time in milliseconds to wait before executing the given {@code computation}.
+     * @param <T>             Type of the computation result.
+     * @return A new Future instance.
+     * @throws NullPointerException if one of executorService of computation is null.
+     */
+    static <T> Future<T> of(ExecutorService executorService, CheckedSupplier<? extends T> computation, long delay) {
         Objects.requireNonNull(executorService, "executorService is null");
         Objects.requireNonNull(computation, "computation is null");
         final FutureImpl<T> future = new FutureImpl<>(executorService);
-        future.run(computation);
+        final CheckedSupplier<? extends T> task = (delay <= 0) ? computation : () -> {
+            Thread.sleep(delay);
```

Comment:
Yes, but Netty does not implement the timeout/delay functionality at the core using delayed task queues.

It looks like to me that Netty uses a ScheduledExecutorService to produce a ScheduledFuture (like I stated above).

These ScheduledFutures are put on a PriorityQueue to ensure that the next delay that expires will be taken in to account. (The underlying java.util.PriorityQueue [does take the natural order of elements into account](https://docs.oracle.com/javase/7/docs/api/java/util/PriorityQueue.html). The java.util.concurrent.ScheduledFuture does extend `Comparable<Delayed>`, by which is ensured that the smaller delays are ordered first.)

**Question:** How uses Netty the PriorityQueue? What is it needed for? The ScheduledFuture will run after a specific delay, regardless if it is put on a queue or not.

Need to take a look at the source code of Netty. But my 'feeling' says, for now we don't need to put our futures on a PriorityQueue!

However, it could be useful to let Javaslang's Future implement `Comparable<java.uti.concurrent.Delayed>`, where the default delay is 0 (for non-delayed Futures).